File: 06_Searching/src/Capacity_To_Ship_Packages_Within_D_Days_1011.java
public class Capacity_To_Ship_Packages_Within_D_Days_1011 {
    public static int shipWithinDays(int[] weights, int days) {
        int start = weights[0];
        int end = 0;

        for (int wei : weights) {
            start = Math.max(wei,start);
            end += wei;
        }

        while (start < end) {
            int mid = start + (end - start) /2;

            if (find(weights,mid,days)) {
                end = mid;
            } else {
                start = mid + 1;
            }
        }

        return end;
    }

    public static boolean find (int[] arr, int mid, int target) {
        int days = 0;
        int cSum = 0;

        for (int wei : arr) {
            if (cSum+wei > mid) {
                cSum = wei;
                days++;
            } else {
                cSum += wei;
            }
        }

        return days < target;
    }

    public static void main(String[] args) {
        System.out.println(shipWithinDays(new int[] {1,2,3,4,5,6,7,8,9,10},5));
    }
}

File: 06_Searching/src/Capacity_To_Ship_Packages_Within_D_Days_1011.js
/**
 * @param {number[]} weights
 * @param {number} days
 * @return {number}
 */
const shipWithinDays = function(weights, days) {
    let start = weights[0];
    let end = 0;

    for (let wei of weights) {
        start = Math.max(start,wei);
        end += wei;
    }


    while (start < end) {
        let mid = Math.floor(start + (end-start)/2);

        if (shipSearch (weights, mid, days)) {
            end = mid;
        } else {
            start = mid + 1;
        }
    }

    return start;
};

function shipSearch (weights, mid, targetDays) {
    let days = 0;
    cSum = 0

    for (let wei of weights) {

        if (cSum+wei > mid) {
            days++;
            cSum = wei;
        } else {
            cSum += wei;
        }
    }

    return days < targetDays;
}

console.log(shipWithinDays([1,2,3,4,5,6,7,8,9,10],5))